static/app/views/explore/logs/useVirtualStreaming.tsx

```diff
@@ -0,0 +1,235 @@
+ *    │   10s    15s    20s    25s    30s  HIDDEN HIDDEN HIDDEN     │
+ *    └─────────────────────────────────────────────────────────────┘
+ *
+ */
+export function useVirtualStreaming(
+  data: InfiniteData<ApiResult<EventsLogsResult>> | undefined
+) {
+  const autoRefresh = useLogsAutoRefresh();
+  const refreshInterval = useLogsRefreshInterval();
+  const rafOn = useRef(false);
+  const [virtualTimestamp, setVirtualTimestamp] = useState<number | undefined>(undefined);
+
+  // If we've received data, initialize the virtual timestamp to be refreshEvery seconds before the max ingest delay timestamp
+  const initializeVirtualTimestamp = useCallback(() => {
+    if (!data?.pages?.length || virtualTimestamp !== undefined) {
+      return;
+    }
+
+    const firstPageWithData = data.pages.find(page => page?.[0]?.data?.length > 0);
+    const pageData = firstPageWithData?.[0]?.data;
+
+    if (!pageData || pageData.length === 0) {
+      return;
+    }
+
+    // Calculate the target timestamp: refreshEvery seconds before the max ingest delay
+    const targetTimestamp = Date.now() - MAX_LOG_INGEST_DELAY - refreshInterval;
+
+    // Find the first row with timestamp less than targetTimestamp
+    let selectedRow = pageData[0];
+    for (const row of pageData) {
+      const rowTimestamp = Number(
+        BigInt(row[OurLogKnownFieldKey.TIMESTAMP_PRECISE]) / 1_000_000n
+      );
+      if (rowTimestamp <= targetTimestamp) {
+        selectedRow = row;
+        break;
+      }
+    }
+
+    const initialTimestamp = selectedRow
+      ? Number(BigInt(selectedRow[OurLogKnownFieldKey.TIMESTAMP_PRECISE]) / 1_000_000n)
+      : Number(
+          BigInt(pageData[0]?.[OurLogKnownFieldKey.TIMESTAMP_PRECISE] ?? 0n) / 1_000_000n
+        );
+
+    setVirtualTimestamp(initialTimestamp);
+  }, [data, virtualTimestamp, refreshInterval]);
+
+  // Initialize when auto refresh is enabled and we have data
+  useEffect(() => {
+    if (autoRefresh && virtualTimestamp === undefined) {
+      initializeVirtualTimestamp();
+    }
+  }, [autoRefresh, initializeVirtualTimestamp, virtualTimestamp]);
+
+  // Reset when auto refresh is disabled
+  useEffect(() => {
+    if (!autoRefresh) {
+      setVirtualTimestamp(undefined);
+    }
+  }, [autoRefresh]);
+
+  // Get the newest timestamp from the latest page to calculate how far behind we are
+  const getMostRecentPageDataTimestamp = useCallback(() => {
+    if (!data?.pages?.length) {
+      return 0;
+    }
+
+    const latestPage = data.pages[data.pages.length - 1];
+    const latestPageData = latestPage?.[0]?.data;
+
+    // Since data is always sorted by timestamp descending, the first row (index 0) is the latest timestamp.
+    return Number(
+      BigInt(latestPageData?.[0]?.[OurLogKnownFieldKey.TIMESTAMP_PRECISE] ?? 0n) /
+        1_000_000n
+    );
+  }, [data]);
+
+  // We setup a RAF loop to update the virtual timestamp smoothly to emulate real-time streaming.
```

Comment:
I wonder if we should have a throttled RAF if we detect bad hardware or older browsers somehow. Maybe do 30fps?

Would be nice to have some stats (send logs?) to track metrics on this.